langchain/document_loaders/directory.py

```diff
@@ -71,22 +89,17 @@ def load(self) -> List[Document]:
                 else:
                     raise e
 
-        for i in items:
-            if i.is_file():
-                if _is_visible(i.relative_to(p)) or self.load_hidden:
-                    try:
-                        sub_docs = self.loader_cls(str(i), **self.loader_kwargs).load()
-                        docs.extend(sub_docs)
-                    except Exception as e:
-                        if self.silent_errors:
-                            logger.warning(e)
-                        else:
-                            raise e
-                    finally:
-                        if pbar:
-                            pbar.update(1)
+        if self.use_multithreading:
+            with concurrent.futures.ThreadPoolExecutor() as executor:
+                executor.map(lambda i: self.load_file(i, p, docs, pbar), items)
```

Comment:
I like the idea of multithreading as it’s a good way to speed up. What I’m not sure are several potential conflicts
- Global Interpreter Lock (GIL) only allows 1 thread at a time. For it, maybe multi-processing is preferred.
- Resource usage is increased. I don’t know where users would use LangChain, when the CPU resource is limited, this can be a concern. They can choose not to use it thorough.

Overall, I think it’s a strategical question to LangChain of whether to support multi-threading/processing. Will wait for maintainers to answer as they have a better view of the big picture.

Reply:
Thank you for the comment!

- Yes I agree about GIL. We have one thread running but there is an exception: waiting for I/O operations. When one thread waits for file to be loaded other can operate. This is the main usage of the concurrent.futures.ThreadPoolExecutor.

- Multithreading is optional and will probably by used only by people who will discover the need. As the one person who requested this originaly.